Update tests to use k3po nukleus transport (not yet complete)

cmebarrow opened the pull request:
Each test case consists of two pairs of scripts, one for high level ("cooked" or decoded form) and one for low level ("raw" or encoded form). The raw scripts are in /nukleus-http.spec/src/main/scripts/org/reaktivity/specification/http for now, my plan is eventually this will be a separate new http.spec project (amounting to a revised version of the existing k3po http specification project). Cooked scripts are in /nukleus-http.spec/src/main/scripts/org/reaktivity/specification/nukleus/http/streams and replace the old agrona scripts.

This PR is not yet ready to merge. Only the architecture tests (raw and cooked) have been completed.

Commit:
Added scripts and an IT illustrating how we plan to use the k3po nukleus streams transport.

cmebarrow commented:
@jfallows Obviously this is not intended for merging, just for review to solidify how we will use the new k3po nukleus streams transport.

Commit:
Added a missing newline

cmebarrow commented on src/test/java/org/reaktivity/specification/nukleus/http/streams/server/rfc7230/newsystem/ArchitectureIT.java:
```diff
@@ -0,0 +1,71 @@
+
+    private final TestRule timeout = new DisableOnDebug(new Timeout(5, SECONDS));
+
+    private final NukleusRule nukleus = new NukleusRule()
```
This will will be changing, probably to be different for the raw and cooked cases.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/http/streams/rfc7230/architecture/request.and.response/cooked/client.rpt:
```diff
@@ -0,0 +1,37 @@
+    option transmission "duplex"
+connected
+
+write extension [0x00] [0x07] ":scheme" [0x0x] "http"
```
might be helpful to support syntax like the following to avoid having to put in the string lengths:
```
write extension [0x00] string ":scheme" string "http"
```

Commit:
Fixed the nukleus rule in the new IT.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/http/streams/rfc7230/architecture/request.and.response/raw/client.rpt:
```diff
@@ -0,0 +1,38 @@
+    nukleus:option source "source"
+    nukleus:option route ${newSourceInputRef}
+    nukleus:option window 8192
+    nukleus:option transmission "duplex"
```
Ideally it would be possible to parameterize the transport URI and options, something like:
```
connect ${connectURI} ${connectOptions}
```
so the script can be used against any HTTP server, for example cutting WebSocket gateway.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/http/streams/rfc7230/architecture/request.and.response/raw/server.rpt:
```diff
@@ -0,0 +1,35 @@
+       option nukleus:source ${source}
+       option nukleus:route ${newTargetOutputRef}
+       option nukleus:window 8192
+       option nukleus:transmission "duplex"
```
I am not sure if the nukleus: options will automatically be applied as options to the http:transport.

Ideally it would be possible to parameterize the transport URI and options, something like:
```
accept http://localhost:8080/ option ${transportURI} ${transportOptions}
```
so the script can be used against any HTTP client (i.e. tcp transport instead of nuklei)

Commit:
Merge branch 'develop' into update_k3po_nukleus_transport

jfallows commented on src/test/java/org/reaktivity/specification/nukleus/http/streams/server/rfc7230/newsystem/ArchitectureIT.java:
```diff
@@ -0,0 +1,68 @@
+    public void shouldCorrelateRequestAndResponseCooked() throws Exception
+    {
+        k3po.start();
+        k3po.finish();
```
just need `k3po.finish()`, right?

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/http/streams/rfc7230/architecture/request.and.response/raw/client.rpt:
```diff
@@ -0,0 +1,38 @@
+    nukleus:option source "source"
+    nukleus:option route ${newSourceInputRef}
+    nukleus:option window 8192
+    nukleus:option transmission "duplex"
```
Ideally it would be possible to parameterize the connect options, something like:
```
connect ${connectURI} ${connectOptions}
```
so the script can be used against any HTTP server, for example Kaazing WebSocket Gateway.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/http/streams/rfc7230/architecture/request.and.response/cooked/client.rpt:
```diff
@@ -0,0 +1,36 @@
+write nukleus:extension [0x00] [0x07] ":scheme" [0x0x] "http"
+write nukleus:extension [0x00] [0x07] ":method" [0x03] "GET"
+write nukleus:extension [0x00] [0x05] ":path" [0x01] "/"
+write nukleus:extension [0x00] [0x0a] ":authority" [0x0e] "localhost:8080"
```
might be helpful to support syntax like the following to avoid having to put in the string lengths:
```
write nukleus:extension [0x00] string ":scheme" string "http"
```

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/http/streams/rfc7230/architecture/request.and.response/raw/server.rpt:
```diff
@@ -0,0 +1,35 @@
+       option nukleus:source ${source}
+       option nukleus:route ${newTargetOutputRef}
+       option nukleus:window 8192
+       option nukleus:transmission "duplex"
```
I am not sure if the nukleus: options will automatically be applied as options to the http:transport.

jfallows replied:
Good point.

I think it might be better to generally handle `options` rather than special-casing the concept of transport.

For example:
```
property httpOptions { "http:transport":"nukleus://streams/target", "nukleus:source": ... }
accept "http://localhost:8080/" options ${httpOptions}
```

Actually, with that in place, it might be better to write the original one as follows:

```
property httpOptions {} // empty

accept "http://localhost:8080/" options ${httpOptions}
```

then use a property override to retarget the transport and options all at once.

Commit:
Updated the proposed scripts (in cooked and raw subdirectories in rfc7230/architecture/request.and.response) to match the latest proposed syntax in nukleus-k3po-ext.java).

Commit:
Updated the proposed scripts to quote URI's.

Commit:
Applied review comments: made HTTP options configurable. Raw scripts now have no knowledge of nukleus transport.

Commit:
Merge branch 'multiple_requests_client' into update_k3po_nukleus_transport

Commit:
Upgraded to k3po 3.0.0-alpha-69. Added dependency on k3po-nukleus-ext.

Commit:
Upgraded to k3po 3.0.0-alpha-69. Added dependency on k3po-nukleus-ext. Incorporated the scripts from k3po http specification, revising the test case names for architecture.

Commit:
Renamed the new http spec scripts to follow the convention client.rpt + server.rpt instead of request.rpt + response.rpt.

Commit:
Renamed some remaining new http spec scripts to follow the convention client.rpt + server.rpt instead of request.rpt + response.rpt.

Commit:
Changed k3po dependency to develop-SNAPSHOT to pick up fix for http transport options.

Commit:
Merge branch 'develop' into update_k3po_nukleus_transport

Commit:
Completed the high and low level ArchitectureIT's and test scripts.

Commit:
Completed the high and low level ArchitectureIT's and test scripts.

Commit:
Merge branch 'develop' of https://github.com/reaktivity/nukleus-http.spec into update_k3po_nukleus_transport

Commit:
Updated licenses in all scripts copied over from k3po/specification/http.

Commit:
Added some flow control tests using the k3po nukleus transport. Updated nukleus.spec dependency to pick up a fix to NukleusRule to avoid keeping a lock on streams files. Added architecture test "response.version.missing".

Commit:
Added http spec version of ConnectionManagementIT and some scripts.

Commit:
Change http spec script /response.with.connection.close/client.rpt to stop using http transport so read closed gets picked up.

Commit:
Fleshed out http spec version of ConnectionManagementIT and scripts. Changed pom to pick up a fix to k3po-nukleus-ext to handle disconnectRequested so abort works for http over nukleus transport.

Commit:
Changed http spec test multiple.requests.pipelined.with.retry to verify the first retried request is not pipelined, which was the original intent of the test. The k3po-nukleus-ext fix is no longer required.

Commit:
http spec test first.pipelined.response.has.connection.close.

Commit:
Completed the upgrade http spec tests.

Commit:
Added many high level connection.management tests. Moved the old agrona transport based scripts into agrona subdirectories.

Commit:
Completed the connection.management tests by adding the high-level proxy tests.

Commit:
Changed multiple.requests.same.connection to verify server side is same connection. Added multiple.requests.different.connections.

Commit:
Completed all message.format http spec tests, except the 4 proxy ones which are ignored.

Commit:
Fixed the architecture spec tests to allow them to be used to test http nukleus as client and server:
- added connect await OUTUT_OUTPUT to the high level client scripts
- changed the low level server scripts to define and use property newTargetOutputRef (as set by the controller route output new script) as the routing ref, defaulting to the value of newSourceInputRef as set by the matching client script (to allow the server and client scripts to run against each other).
- changed the high-level client scripts to use newSourceOutputRef (as set by the controller route output new script)
- changed the high-level server scripts to use newTargetInputRef (as set by the controller route input new script), defaulting to the value of newSourceOutputRefas set by the matching client script (so they can still run against each other).
- Altered request.uri.with.percent.chars to expect the :path to the header to remain encoded in BEGIN

Commit:
Fixed the agrona server scripts to accomodate eary BEGIN on source output established.

Commit:
Fixed the agrona server nukleus scripts to accomodate early BEGIN on source output established.

Commit:
Couple of script corrections.

Commit:
More fixes to test scripts to allow them to be consumered by http nukleus and pass in the spec build.

Commit:
More fixes to test scripts (for content-length)

Commit:
More fixes to test scripts (for content-length)
